cpp/src/arrow/ipc/metadata-internal.cc

```diff
@@ -1143,15 +1150,16 @@ Status GetSparseTensorMetadata(const Buffer& metadata, std::shared_ptr<DataType>
                                std::vector<std::string>* dim_names,
                                int64_t* non_zero_length,
                                SparseTensorFormat::type* sparse_tensor_format_id) {
-  auto message = flatbuf::GetMessage(metadata.data());
-  if (message->header_type() != flatbuf::MessageHeader_SparseTensor) {
-    return Status::IOError("Header of flatbuffer-encoded Message is not SparseTensor.");
+  flatbuffers::Verifier verifier(metadata.data(), metadata.size(), 128);
+  if (!flatbuf::VerifyMessageBuffer(verifier)) {
+    return Status::IOError("Verification of flatbuffer-encoded Message failed.");
   }
-  if (message->header() == nullptr) {
-    return Status::IOError("Header-pointer of flatbuffer-encoded Message is null.");
+  auto message = flatbuf::GetMessage(metadata.data());
+  auto sparse_tensor = message->header_as_SparseTensor();
+  if (sparse_tensor == nullptr) {
+    return Status::IOError(
+        "Header-type of flatbuffer-encoded Message is not SparseTensor.");
   }
```

Comment:
This assertion about the message type could be handled more generically (since we can pass in the expected union value), then this helps with code duplication

Reply:
And I have the same issue here: how to pull that code out that it includes the assignment AND the return?